src/tezos_interop/run_entrypoint.ml

```diff
@@ -2,63 +2,217 @@ open Helpers
+      type t = {
+        id : id;
+        rpc_node : string;
+        secret : string;
+        confirmation : int;
+        destination : string;
+        entrypoint : string;
+        payload : Yojson.Safe.t;
+      }
+      [@@deriving to_yojson]
+    end in
+    let { id; content } = request in
+    let { rpc_node; secret; confirmation; destination; entrypoint; payload } =
+      content in
+    T.to_yojson
+      { id; rpc_node; secret; confirmation; destination; entrypoint; payload }
+end
+
+module Response = struct
+  type transaction =
+    | Applied     of { hash : string }
+    | Failed      of { hash : string }
+    | Skipped     of { hash : string }
+    | Backtracked of { hash : string }
+    | Unknown     of { hash : string }
+    | Error       of string
+  type t = {
+    id : id;
+    content : transaction;
+  }
+
+  let transaction_of_yojson json =
+    let module T = struct
+      type t = { status : string }
+      and finished = { hash : string }
+      and error = { error : string } [@@deriving of_yojson { strict = false }]
+    end in
+    let finished make =
+      let%ok { hash } = T.finished_of_yojson json in
+      Ok (make hash) in
+    let%ok { status } = T.of_yojson json in
+    match status with
+    | "applied" -> finished (fun hash -> Applied { hash })
+    | "failed" -> finished (fun hash -> Failed { hash })
+    | "skipped" -> finished (fun hash -> Skipped { hash })
+    | "backtracked" -> finished (fun hash -> Backtracked { hash })
+    | "unknown" -> finished (fun hash -> Unknown { hash })
+    | "error" ->
+      let%ok { error } = T.error_of_yojson json in
+      Ok (Error error)
+    | _ -> Error "invalid status"
+
+  let of_yojson json =
+    let module T = struct
+      type header = { id : int } [@@deriving of_yojson { strict = false }]
+    end in
+    let%ok { id } = T.header_of_yojson json in
+    let%ok content = transaction_of_yojson json in
+    Ok { id; content }
+end
+
+module Process : sig
+  type t
+  val spawn : unit -> t
+  val request : t -> Request.transaction -> Response.transaction Lwt.t
+end = struct
+  exception Failed_to_parse_json of string
+  exception Process_closed of Unix.process_status
+  exception Unknown_responde_id of Response.t
+
+  type t = {
+    mutable next_id : id;
+    mutable request_promise : Request.t Lwt.t;
+    mutable request_resolver : Request.t Lwt.u;
+    mutable response_pending : Response.transaction Lwt.u Id_map.t;
+  }
+
+  let make () =
+    let request_promise, request_resolver = Lwt.wait () in
+    {
+      next_id = initial_id;
+      request_promise;
+      request_resolver;
+      response_pending = Id_map.empty;
+    }
+
+  let on_fail exn =
+    Format.eprintf "tezos_interop failure: %s\n%!" (Printexc.to_string exn);
+    exit 1
+
+  let spawn_node_process ~file ~to_yojson ~of_yojson input_stream =
+    let node = "node" in
+    let command = (node, [|node; file|]) in
+
+    let process = Lwt_process.open_process command in
```

Comment:
LWT is still black magic to me. I can't find [any real documentation for this function](https://dsheets.github.io/codoc/lwt.2.4.6/_build/src/unix/lwt_process/index.html#/val:open_process) so I'm not sure I can review this section of the code properly

Reply:
Even if there was documentation it would not document the behavior properly, this requires a LOT of Unix knowledge to understand in details, which I don't have. It's one of the cases where even if you look at the implementation is not enough and I did.